pkg/plugin/sdk/deployment_source.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2025 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdk
+
+import (
+	"encoding/json"
+	"fmt"
+
+	config "github.com/pipe-cd/pipecd/pkg/configv1"
+	"github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1/common"
+)
+
+// DeploymentSource represents the source of the deployment.
+type DeploymentSource[Spec any] struct {
+	// ApplicationDirectory is the directory where the source code is located.
+	ApplicationDirectory string
+	// CommitHash is the git commit hash of the source code.
+	CommitHash string
+	// ApplicationConfig is the configuration of the application.
+	ApplicationConfig *ApplicationConfig[Spec]
+	// ApplicationConfigFilename is the name of the file that contains the application configuration.
+	// The plugins can use this to avoid mistakenly reading this file as a manifest to deploy.
+	ApplicationConfigFilename string
+}
+
+// newDeploymentSource converts the common.DeploymentSource to the internal representation.
+func newDeploymentSource[Spec any](source *common.DeploymentSource) (DeploymentSource[Spec], error) {
+	cfg, err := config.DecodeYAML[*ApplicationConfig[Spec]](source.GetApplicationConfig())
+	if err != nil {
+		return DeploymentSource[Spec]{}, fmt.Errorf("failed to decode application config: %w", err)
+	}
+	return DeploymentSource[Spec]{
+		ApplicationDirectory:      source.GetApplicationDirectory(),
+		CommitHash:                source.GetCommitHash(),
+		ApplicationConfig:         cfg.Spec,
+		ApplicationConfigFilename: source.GetApplicationConfigFilename(),
+	}, nil
+}
+
+// DecodeYAML decodes the application config from YAML data.
+// This is intended to use in the tests.
+func DecodeYAML[Spec any](data []byte) (*ApplicationConfig[Spec], error) {
+	cfg, err := config.DecodeYAML[*ApplicationConfig[Spec]](data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode application config: %w", err)
+	}
+	return cfg.Spec, nil
+}
+
+type ApplicationConfig[Spec any] struct {
+	commonSpec *config.GenericApplicationSpec
+	Spec       *Spec
+}
```

Comment:
[nit] plz add comment for the struct and fields 🙏

Reply:
Added comments
https://github.com/pipe-cd/pipecd/pull/5740/commits/c92050ddaff4d5cdae692e1e49d4334ace94ef93